### ISSUE TITLE
NoClassDefFoundError: Reika/ChromatiCraft/ModInterface/ThaumCraft/ItemManipulatorFocus

### ISSUE THREAD

**Issue by CaiB** [ReikaKalseki/Reika_Mods_Issues#2818]
Hello!

Using a clean 1.7.10 install, and adding your mods works fine. However the moment I install CoFH Core (using the newest version available for 1.7.10), ChromatiCraft v29g crashes on startup. I do not plan to install ThaumCraft.

Mods:
```
UC	mcp{9.05} [Minecraft Coder Pack] (minecraft.jar) 
UC	FML{7.10.99.99} [Forge Mod Loader] (forge-1.7.10-10.13.4.1614-1.7.10-universal.jar) 
UC	Forge{10.13.4.1614} [Minecraft Forge] (forge-1.7.10-10.13.4.1614-1.7.10-universal.jar) 
UC	<CoFH ASM>{000} [CoFH ASM] (minecraft.jar) 
UC	<DragonAPI ASM>{0} [DragonAPI ASM Data Initialization] (minecraft.jar) 
UC	CoFHCore{1.7.10R3.1.4} [CoFH Core] (1.7.10 - FW CoFHCore 3.1.4-329.jar) 
UC	DragonAPI{v29g} [DragonAPI] (1.7.10 - FW DragonAPI V29g.jar) 
UE	ChromatiCraft{v29g} [ChromatiCraft] (1.7.10 - ChromatiCraft V29g.jar) 
UC	RotaryCraft{v29g} [RotaryCraft] (1.7.10 - RotaryCraft V29g.jar) 
UC	ElectriCraft{v29a} [ElectriCraft] (1.7.10 - ElectriCraft V29a.jar) 
UC	ExpandedRedstone{v29b} [ExpandedRedstone] (1.7.10 - ExpandedRedstone V29b.jar) 
UC	ReactorCraft{v29e} [ReactorCraft] (1.7.10 - ReactorCraft V29e.jar) 
```

Error:
```
java.lang.NoClassDefFoundError: Reika/ChromatiCraft/ModInterface/ThaumCraft/ItemManipulatorFocus
	at Reika.ChromatiCraft.Registry.ChromaItems.<clinit>(ChromaItems.java:199) ~[ChromaItems.class:?]
	at Reika.ChromatiCraft.ChromatiCraft.<clinit>(ChromatiCraft.java:265) ~[ChromatiCraft.class:?]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_275]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_275]
	at cpw.mods.fml.common.FMLModContainer.constructMod(FMLModContainer.java:440) ~[forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_275]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_275]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_275]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_275]
	at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.dispatch(EventBus.java:322) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.post(EventBus.java:275) ~[guava-17.0.jar:?]
	at cpw.mods.fml.common.LoadController.sendEventToModContainer(LoadController.java:212) ~[forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at cpw.mods.fml.common.LoadController.propogateStateMessage(LoadController.java:190) ~[forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_275]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_275]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_275]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_275]
	at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.dispatch(EventBus.java:322) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304) ~[guava-17.0.jar:?]
	at com.google.common.eventbus.EventBus.post(EventBus.java:275) ~[guava-17.0.jar:?]
	at cpw.mods.fml.common.LoadController.distributeStateMessage(LoadController.java:119) [LoadController.class:?]
	at cpw.mods.fml.common.Loader.loadMods(Loader.java:513) [Loader.class:?]
	at cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:208) [FMLClientHandler.class:?]
	at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:480) [bao.class:?]
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:878) [bao.class:?]
	at net.minecraft.client.main.Main.main(SourceFile:148) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_275]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_275]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_275]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_275]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_275]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_275]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_275]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_275]
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196) [NewLaunch.jar:?]
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.listen(EntryPoint.java:143) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.main(EntryPoint.java:34) [NewLaunch.jar:?]
Caused by: java.lang.ClassNotFoundException: Reika.ChromatiCraft.ModInterface.ThaumCraft.ItemManipulatorFocus
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_275]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_275]
	... 45 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: thaumcraft.api.wands.ItemFocusBasic
	at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-17.0.jar:?]
	at cofh.asm.CoFHAccessTransformer.transform(CoFHAccessTransformer.java:94) ~[1.7.10%20-%20FW%20CoFHCore%203.1.4-329.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_275]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_275]
	... 45 more
Caused by: java.lang.ClassNotFoundException: thaumcraft.api.wands.ItemFocusBasic
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:101) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_275]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_275]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_275]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_275]
	at cofh.asm.CoFHAccessTransformer.transform(CoFHAccessTransformer.java:90) ~[1.7.10%20-%20FW%20CoFHCore%203.1.4-329.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_275]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_275]
	... 45 more
```

The full log is attached here:
[fml-client-latest.log](https://github.com/ReikaKalseki/Reika_Mods_Issues/files/6300537/fml-client-latest.log)

**Comment by CaiB** [comment]
Not sure why I didn't find #2558 when I searched. It's not super clear what the resolution to this is however, as there is no actual information given, each issue just has a reply containing only a list of links to other issues that are closed with lists of links to other issues.

If I understood leldra2 correctly, the resolution should be to delete the file `cofh/api/modhelpers/ThaumcraftHelper.class` from inside the CoFH Core JAR file. However, when I do this, the error persists. I've even gone as far as deleting the entire `cofh/api` folder and it's still the same error and crash.

**Comment by CaiB** [comment]
This works fine if I use ChromatiCraft v24c + DragonAPI v24c + the same CoFH core. Since old versions of your mods are not available, I'm unable to track down when this issue started.

**Comment by MalTeeez** [comment]
After reading through #2558 I have to ask if you tried deleting the TC API helper class in your cofh-core jar

**Comment by CaiB** [comment]
> After reading through #2558 I have to ask if you tried deleting the TC API helper class in your cofh-core jar

Please see my reply above, where I said
> If I understood leldra2 correctly, the resolution should be to delete the file `cofh/api/modhelpers/ThaumcraftHelper.class` from inside the CoFH Core JAR file. However, when I do this, the error persists. I've even gone as far as deleting the entire `cofh/api` folder and it's still the same error and crash.

Is this the file you meant? If so, that didn't work for me. If not, please let me know if there is another file I should be deleting instead.

**Comment by MalTeeez** [comment]
Sorry, no idea how I was able to miss that part. 
I tried tinkering around with it a bit too but seems like this'll have to be fixed in code.

**Comment by ReikaKalseki** [comment]
> If I understood leldra2 correctly, the resolution should be to delete the file cofh/api/modhelpers/ThaumcraftHelper.class from inside the CoFH Core JAR file.

No. You need to remove their copy of the TC api.

**Comment by CaiB** [comment]
@ReikaKalseki Please point out where in the JAR I can find that. For reference, here's the folder structure of the [unmodified CoFH Core JAR](https://www.curseforge.com/minecraft/mc-mods/cofh-core/files/2388750):
```
CoFHCore-[1.7.10]3.1.4-329.jar
├─assets
│  └─cofh
│      ├─lang
│      ├─textures
│      │  ├─gui
│      │  │  ├─elements
│      │  │  └─storage
│      │  └─items
│      │      └─icons
│      └─world
├─cofh
│  ├─api
│  │  ├─block
│  │  ├─core
│  │  ├─energy
│  │  ├─fluid
│  │  ├─inventory
│  │  ├─item
│  │  ├─modhelpers
│  │  ├─tileentity
│  │  ├─transport
│  │  └─world
│  ├─asm
│  │  └─relauncher
│  ├─asmhooks
│  │  ├─block
│  │  └─event
│  ├─core
│  │  ├─augments
│  │  ├─block
│  │  ├─chat
│  │  ├─command
│  │  ├─crash
│  │  ├─enchantment
│  │  ├─entity
│  │  ├─fluid
│  │  ├─gui
│  │  │  ├─client
│  │  │  ├─container
│  │  │  ├─element
│  │  │  └─slot
│  │  ├─item
│  │  │  └─tool
│  │  ├─key
│  │  ├─network
│  │  ├─particles
│  │  ├─render
│  │  │  ├─customcharrendering
│  │  │  └─hitbox
│  │  ├─sided
│  │  ├─util
│  │  │  ├─crafting
│  │  │  ├─energy
│  │  │  ├─fluid
│  │  │  ├─nbt
│  │  │  └─oredict
│  │  └─world
│  │      ├─decoration
│  │      └─feature
│  ├─lib
│  │  ├─audio
│  │  ├─gui
│  │  │  ├─container
│  │  │  ├─element
│  │  │  │  └─listbox
│  │  │  └─slot
│  │  ├─inventory
│  │  ├─network
│  │  ├─render
│  │  │  └─particle
│  │  ├─transport
│  │  ├─util
│  │  │  ├─helpers
│  │  │  └─position
│  │  └─world
│  │      ├─biome
│  │      └─feature
│  ├─mod
│  │  └─updater
│  ├─plugins
│  │  └─nei
│  └─repack
│      ├─codechicken
│      │  └─lib
│      │      ├─asm
│      │      ├─colour
│      │      ├─config
│      │      ├─lighting
│      │      ├─math
│      │      ├─raytracer
│      │      ├─render
│      │      │  └─uv
│      │      ├─util
│      │      └─vec
│      └─immibis
│          └─bon
│              └─mcp
├─META-INF
└─skyboy
    └─core
        └─world
```
If it helps, here's that same structure listing but with all of the files shown as well: https://pastebin.com/634wPPmK

**Comment by ReikaKalseki** [comment]
OK, maybe it is in another jar (one of the other CoFH jars, perhaps). Classloading is messy this way.

**Comment by CaiB** [comment]
@ReikaKalseki I specifically have only the CoFH Core JAR installed, as I narrowed it down to causing the issue. Also, like said, I am able to get it working by deleting the file inside the JAR when I use ChromatiCraft v24c + DragonAPI v24c. There's something in newer versions of your mods that is causing this issue to come up.

```
.minecraft
│  icon.png
│
├─asm
│  └─ccl_modular
│          aji.txt
│          alb.txt
│
├─config
│  │  Baubles.cfg
│  │  CodeChickenCore.cfg
│  │  forge.cfg
│  │  forgeChunkLoading.cfg
│  │  microblocks.cfg
│  │  MrTJPCore.cfg
│  │  multipart.cfg
│  │  OpenModsLibCore.json
│  │  splash.properties
│  │
│  ├─cofh
│  │  ├─core
│  │  │      client.cfg
│  │  │      common.cfg
│  │  │      friends.cfg
│  │  │      loot.cfg
│  │  │
│  │  └─world
│  │          Vanilla.json
│  │
│  └─Reika
│          ChromatiCraft.cfg
│          CHROMATICRAFT_PackModifications.cfg
│          DragonAPI.cfg
│          DragonAPI_Aux_Retrogen.cfg
│          ElectriCraft.cfg
│          ELECTRICRAFT_PackModifications.cfg
│          ExpandedRedstone.cfg
│          ReactorCraft.cfg
│          REACTORCRAFT_PackModifications.cfg
│          RotaryCraft.cfg
│          RotaryCraft_CustomExtracts.cfg
│          RotaryCraft_CustomGPRColors.cfg
│          ROTARYCRAFT_PackModifications.cfg
│
├─coremods
├─FailedClasses
│      Reika.ChromatiCraft.ModInterface.Bees.TileEntityLumenAlveary.classdata
│      Reika.DragonAPI.ModInteract.DeepInteract.FrameBlacklist.classdata
│      Reika.DragonAPI.ModInteract.DeepInteract.ReikaMystcraftHelper.classdata
│
├─mods
│  │  ChromatiCraft+1.7.10+V29g.jar
│  │  CoFHCore-[1.7.10]3.1.4-329.jar
│  │  DragonAPI+1.7.10+V29g.jar
│  │  ElectriCraft+1.7.10+V29a.jar
│  │  ExpandedRedstone+1.7.10+V29b.jar
│  │  ReactorCraft+1.7.10+V29e.jar
│  │  RotaryCraft+1.7.10+V29g.jar
│  │
│  └─Reika
│      └─ChromatiCraft
│          └─Music
│                  elemental precursors.ogg
│                  ethereal contact_c.ogg
│                  explorer.ogg
│                  file_list.dat
│                  glowing horizons.ogg
│                  luminescence.ogg
│                  nightfall.ogg
│                  polychromacy_c.ogg
│                  singularity.ogg
│                  transcendental journey.ogg
│
├─resourcepacks
├─saves
│  └─DragonAPI
├─screenshots
├─server-resource-packs
└─texturepacks
```

**Comment by CaiB** [comment]
For the record, deleting everything else in the .minecraft folder except the `mods` folder containing nothing but CoFH Core + DragonAPI + ChromatiCraft still causes the crash. I really don't know what more I can really say here to get you to actually look into it.

```
.minecraft
│  icon.png
│
└─mods
        ChromatiCraft+1.7.10+V29g.jar
        CoFHCore-[1.7.10]3.1.4-329.jar
        DragonAPI+1.7.10+V29g.jar
```

Launching it gets
```
[21:00:13] [Client thread/ERROR] [FML]: The following problems were captured during this phase
[21:00:13] [Client thread/ERROR] [FML]: Caught exception from ChromatiCraft
java.lang.NoClassDefFoundError: Reika/ChromatiCraft/ModInterface/ThaumCraft/ItemManipulatorFocus
	at Reika.ChromatiCraft.Registry.ChromaItems.<clinit>(ChromaItems.java:199) ~[ChromaItems.class:?]
	at Reika.ChromatiCraft.ChromatiCraft.<clinit>(ChromatiCraft.java:265) ~[ChromatiCraft.class:?]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_275]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_275]
	...
```

**Comment by ReikaKalseki** [comment]
If removing CoFH fixes it, then they are doing something with classloading that breaks my ability to "reparent" that class.

**Comment by CaiB** [comment]
Since you seem completely unwilling to look into, let alone fix this defect in recent versions of your mods, I've gone ahead and found a workaround.

If anyone else encounters the problem, just grab the newest ThaumCraft JAR for 1.7.10, and delete everything except the `/thaumcraft/api` folder's contents, then toss it into your mods folder. The JAR should now be quite small. (Note that the API doesn't add any of the gameplay content, so you're still playing without ThaumCraft)

Essentially, you just need to do exactly the opposite of what the supposed fix is. Add the API if it's not present so that DragonAPI's broken dependency handling is avoided.

**Comment by ReikaKalseki** [comment]
Yes, adding the missing API is going to fix missing API issues. However, as I have repeatedly stated, I have code in place that removes the references to the API as necessary, code that works *the majority of the time*.

**Comment by CaiB** [comment]
Well, clearly this issue is outside of that supposed majority. Took you 3 replies of being entirely dismissive even after I provided ample information to easily reproduce the issue as well as evidence both that I tried to resolve it myself and that it was caused by a change on your part.

Perhaps instead of essentially telling people to leave by immediately closing the issue without resolution and being dismissive, it may help create a better environment if you actually read the issue and took the minute or two to verify something before making inapplicable suggestions. Probably would've saved time overall too.
